### PR TITLE
Feat/stack nopanic

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

addresses #138 while not completely resolving it.
splitting this issue into 3 PRs for clarity

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

## Added changes to [`CHANGELOG.md`](../CHANGELOG.md)

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/tests/test_stack.cairo [hunk 1]
```diff
@@ -139,32 +149,38 @@ mod pop {
...
         // When & Then
         let result = stack.pop();
-        assert(result.is_none(), 'should return None');
+        assert(result.is_err(), 'should return None');
     }
 
     #[test]
     #[available_gas(50000)]
-    #[should_panic(expected: ('Kakarot: StackUnderflow', ))]
     fn test_pop_n_should_fail_when_stack_underflow() {
```
Comment (enitrat) [inline src/tests/test_stack.cairo:163]:
```suggestion
    fn test_pop_n_should_return_err_when_stack_underflow() {
```

File: src/tests/test_stack.cairo [hunk 2]
```diff
@@ -207,27 +223,35 @@ mod peek {
...
 
         // Then
         assert(result == 2, 'wrong result');
     }
 
     #[test]
-    #[available_gas(35000)]
-    #[should_panic(expected: ('Kakarot: StackUnderflow', ))]
+    #[available_gas(350000)]
     fn test_should_fail_when_underflow() {
```
Comment (enitrat) [inline src/tests/test_stack.cairo:234]:
```suggestion
    fn test_should_return_err_when_underflow() {
```

File: src/tests/test_stack.cairo [hunk 3]
```diff
@@ -237,49 +261,59 @@ mod swap {
...
         assert(index1 == 3, 'post-swap: wrong index1');
-        let index0 = stack.peek_at(0);
+        let index0 = stack.peek_at(0).unwrap();
         assert(index0 == 2, 'post-swap: wront index0');
     }
 
     #[test]
     #[available_gas(50000)]
-    #[should_panic(expected: ('Kakarot: StackUnderflow', ))]
     fn test_should_fail_when_index_1_is_underflow() {
```
Comment (enitrat) [inline src/tests/test_stack.cairo:289]:
```suggestion
    fn test_should_return_err_when_index_1_is_underflow() {
```

File: src/tests/test_stack.cairo [hunk 4]
```diff
@@ -237,49 +261,59 @@ mod swap {
...
+            result.unwrap_err() == EVMError::StackError(STACK_UNDERFLOW),
+            'should return StackUnderflow'
+        );
     }
 
     #[test]
-    #[available_gas(60000)]
-    #[should_panic(expected: ('Kakarot: StackUnderflow', ))]
+    #[available_gas(600000)]
     fn test_should_fail_when_index_2_is_underflow() {
```
Comment (enitrat) [inline src/tests/test_stack.cairo:305]:
```suggestion
    fn test_should_return_err_when_index_2_is_underflow() {
```

File: src/stack.cairo [hunk 5]
```diff
@@ -69,31 +69,32 @@ impl StackImpl of StackTrait {
     /// # Panics
     /// If this operation would overflow the stack, 
     /// panics with a StackOverflow error.
```
Comment (enitrat) [inline src/stack.cairo:71]:
```suggestion
    /// returns with a StackOverflow error.
```

File: src/stack.cairo [hunk 6]
```diff
@@ -69,31 +69,32 @@ impl StackImpl of StackTrait {
...
-            panic_with_felt252(errors::STACK_OVERFLOW)
+            return Result::Err(EVMError::StackError(STACK_OVERFLOW));
         }
         self.items.insert(self.len.into(), NullableTraitExt::new(item));
         self.len += 1;
+        Result::Ok(())
     }
 
     /// Pops the top item off the stack. If the stack is empty,
     /// leaves the stack unchanged.
```
Comment (enitrat) [inline src/stack.cairo:83]:
```suggestion
    /// returns with a StackOverflow error.
```

**Comment by enitrat** [comment]
@Eikix can you resolve conflict and rebase on main, then generate the snapshot file and commit it pls ? :)

**Review by enitrat — APPROVED** [review]

**Review by enitrat — APPROVED** [review]
LETS GO

File: src/instructions/comparison_operations.cairo [hunk 7]
```diff
@@ -4,89 +4,108 @@ use kakarot::context::ExecutionContextTrait;
...
-        match maybe_a {
-            Option::Some(a) => {
-                result = ~a;
-            },
-            Option::None => {
-                panic_with_felt252(STACK_UNDERFLOW);
-            }
-        }
+    fn exec_not(ref self: ExecutionContext) -> Result<(), EVMError> {
+        let a = self.stack.pop()?;
```
[the remark this answers is not in the document]
Comment (danilowhk) [reply src/instructions/comparison_operations.cairo:82]: V cool

**Pull request merged**